### PR TITLE
[serve] add support for compact strategy + max replicas per node

### PR DESCRIPTION
**Pull request by zcin**
[serve] add support for compact strategy + max replicas per node

Add support for compact scheduling strategy with `max_replicas_per_node`.

Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

### CODE REVIEW

**Comment by zcin** [comment]
@edoakes @jjyao PTAL!

File: python/ray/serve/_private/deployment_scheduler.py [hunk 1]
```diff
@@ -144,12 +144,25 @@ def required_resources(self) -> Resources:
...
+        # Using implicit resource (resources that every node
+        # implicitly has and total is 1)
+        # to limit the number of replicas on a single node.
+        if self.max_replicas_per_node is not None:
+            deployment_id = self.replica_id.deployment_id
+            implicit_resource = (
+                f"{ray._raylet.IMPLICIT_RESOURCE_PREFIX}"
+                f"{deployment_id.app_name}:{deployment_id.name}"
+            )
+            required[implicit_resource] = 1.0 / self.max_replicas_per_node
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_scheduler.py:163]: What is this magic about? Where is the implicit resource generated?

Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py:163]: This is generated in core, by default all nodes have 1.0 amount of resources that start with a certain prefix. cc: @jjyao

File: python/ray/serve/_private/deployment_scheduler.py [hunk 2]
```diff
@@ -184,9 +199,18 @@ def required_resources(self) -> Resources:
...
         else:
-            return self.actor_resources
+            required = self.actor_resources
+
+        if self.max_replicas_per_node:
+            implicit_resource = (
+                f"{ray._raylet.IMPLICIT_RESOURCE_PREFIX}"
+                f"{self.deployment_id.app_name}:{self.deployment_id.name}"
+            )
+            required[implicit_resource] = 1.0 / self.max_replicas_per_node
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:211]: Move this into a helper function so the code can be reused.

File: python/ray/serve/_private/deployment_scheduler.py [hunk 3]
```diff
@@ -211,6 +235,18 @@ class LaunchingReplicaInfo:
     target_labels: Optional[Dict[str, Any]] = None
 
 
+def _flatten(
+    deployment_to_replicas: Dict[DeploymentID, Dict[str, Any]]
+) -> Dict[ReplicaID, Any]:
+    """Flattens a dict of {deployment_id: {replica_id: val}} to {replica_id: val}."""
+
+    return {
+        replica_id: val
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:244]: just to double confirm that replica_id is globally unique not only unique within a deployment

File: python/ray/serve/_private/deployment_scheduler.py [hunk 4]
```diff
@@ -211,6 +235,18 @@ class LaunchingReplicaInfo:
     target_labels: Optional[Dict[str, Any]] = None
 
 
+def _flatten(
+    deployment_to_replicas: Dict[DeploymentID, Dict[str, Any]]
```
Comment (jjyao) [inline python/ray/serve/_private/deployment_scheduler.py:239]: Dict[DeploymentID, Dict[ReplicaID, Any]] ?

File: python/ray/serve/tests/unit/test_deployment_scheduler.py [hunk 5]
```diff
@@ -788,6 +807,68 @@ def test_heterogeneous_resources(self):
             assert scheduling_strategy.node_id == "node1"
             assert call.kwargs == {"placement_group": None}
 
+    @pytest.mark.parametrize("use_pg", [True, False])
+    def test_max_replicas_per_node(self, use_pg: bool):
```
Comment (jjyao) [inline python/ray/serve/tests/unit/test_deployment_scheduler.py:811]: Could you add a doc string saying what this test is testing?

File: python/ray/serve/tests/unit/test_deployment_scheduler.py [hunk 6]
```diff
@@ -788,6 +807,68 @@ def test_heterogeneous_resources(self):
             assert scheduling_strategy.node_id == "node1"
             assert call.kwargs == {"placement_group": None}
 
+    @pytest.mark.parametrize("use_pg", [True, False])
```
Comment (jjyao) [inline python/ray/serve/tests/unit/test_deployment_scheduler.py:810]: I think currently max_replicas_per_node doesn't work with pg (The semantic is only clear when it's STRICT_PACK).

Let me actually create a PR to throw an error when both max_replicas_per_node and pg are specified.

File: python/ray/serve/_private/deployment_scheduler.py (hunk 3 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py:244]: Yup!

File: python/ray/serve/_private/deployment_scheduler.py (hunk 2 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_scheduler.py:211]: I will make a follow up PR to unify the two data structures, if that's OK.

Comment (jjyao) [reply python/ray/serve/_private/deployment_scheduler.py:211]: sure

File: python/ray/serve/tests/unit/test_deployment_scheduler.py (hunk 6 above)
Comment (zcin) [reply python/ray/serve/tests/unit/test_deployment_scheduler.py:810]: Yeah `use_pg` actually means `use_strict_pack_pg`, let me rename it.

**Review by jjyao — APPROVED** [review]

File: python/ray/serve/tests/unit/test_deployment_scheduler.py (hunk 6 above)
Comment (zcin) [reply python/ray/serve/tests/unit/test_deployment_scheduler.py]: Synced offline, removing this test and making an assumption that max_replicas_per_node and placement_group_bundles will not both be specified and do not work together.

Comment (akshay-anyscale) [reply python/ray/serve/tests/unit/test_deployment_scheduler.py]: will these not work together even if its strict pack?

Comment (edoakes) [reply python/ray/serve/tests/unit/test_deployment_scheduler.py]: Also if this is the case let's make sure it's documented & raises a good error

Comment (zcin) [reply python/ray/serve/tests/unit/test_deployment_scheduler.py]: Yeah I think @jjyao is working on a PR to add validation + throw good error when user tries to specify both.

Comment (akshay-anyscale) [reply python/ray/serve/tests/unit/test_deployment_scheduler.py]: We haven't really documented pg btw and the main use is in rayllm where it's always strict pack

Comment (jjyao) [reply python/ray/serve/tests/unit/test_deployment_scheduler.py]: In theory, max_replicas_per_node can work with strict_pack pg but given rayllm doesn't use max_replicas_per_node, we can add the support later. For now, I'll just create a PR to make these two options mutually exclusive.

**Pull request merged**